File: src/main/java/frc/robot/commands/RealTimeAimDrive.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import static edu.wpi.first.units.Units.*;

import com.ctre.phoenix6.swerve.SwerveModule.DriveRequestType;
import com.ctre.phoenix6.swerve.SwerveModule.SteerRequestType;
import com.ctre.phoenix6.swerve.SwerveRequest;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;

import frc.robot.generated.TunerConstants;
import frc.robot.subsystems.CommandSwerveDrivetrain;

public class RealTimeAimDrive extends Command {
    private final CommandSwerveDrivetrain drive;
    private final double shooterFacingOffsetDeg;
    private final double maxSpeedMetersPerSecond =
            TunerConstants.kSpeedAt12Volts.in(MetersPerSecond);

    private final SwerveRequest.FieldCentricFacingAngle facingRequest =
            new SwerveRequest.FieldCentricFacingAngle();

    public RealTimeAimDrive(CommandSwerveDrivetrain drive) {
        this(drive, 180.0);
    }

    public RealTimeAimDrive(CommandSwerveDrivetrain drive, double shooterFacingOffsetDeg) {
        this.drive = drive;
        this.shooterFacingOffsetDeg = shooterFacingOffsetDeg;
        addRequirements(drive);
    }

    @Override
    public void initialize() {
        drive.setSomAngleDiff(180.0);
        facingRequest
                .withHeadingPID(8.0, 0.0, 0.1)
                .withDeadband(maxSpeedMetersPerSecond * 0.05)
                .withRotationalDeadband(0.1)
                .withMaxAbsRotationalRate(4.71)
                .withDriveRequestType(DriveRequestType.Velocity)
                .withSteerRequestType(SteerRequestType.Position);
    }

    @Override
    public void execute() {
        Pose2d pose = drive.getState().Pose;
        Translation2d hub = drive.getHubPosition();
        Rotation2d targetDirection = Rotation2d
                .fromRadians(Math.atan2(hub.getY() - pose.getY(), hub.getX() - pose.getX()))
                .minus(Rotation2d.fromDegrees(shooterFacingOffsetDeg));
        double angleErrorDeg = Math.abs(targetDirection.minus(pose.getRotation()).getDegrees());

        drive.setSomAngleDiff(angleErrorDeg);
        if (angleErrorDeg < 2.0) {
            drive.setBrakeRequest();
        } else {
            drive.setControl(
                    facingRequest
                            .withVelocityX(0.0)
                            .withVelocityY(0.0)
                            .withTargetDirection(targetDirection));
        }

        SmartDashboard.putNumber("Shooting/AimErrorDeg", angleErrorDeg);
        SmartDashboard.putBoolean("Shooting/DrivetrainLocked", angleErrorDeg < 2.0);
    }

    @Override
    public void end(boolean interrupted) {
        drive.setSomAngleDiff(180.0);
        drive.setSafeCoastWithCurrentSpeeds();
    }

    @Override
    public boolean isFinished() {
        return false;
    }
}

File: cpp/src/main/cpp/commands/RealTimeAimDrive.cpp
#include "commands/RealTimeAimDrive.h"

#include <cmath>

#include <frc/smartdashboard/SmartDashboard.h>

using namespace units::literals;

RealTimeAimDrive::RealTimeAimDrive(
    subsystems::CommandSwerveDrivetrain* drive,
    units::degree_t shooterFacingOffset)
    : drive_(drive), shooter_facing_offset_(shooterFacingOffset) {
  AddRequirements({drive});
}

void RealTimeAimDrive::Initialize() {
  drive_->SOMangleDiff = 180.0;
  facing_request_.WithHeadingPID(8, 0, 0.1)
      .WithDeadband(max_speed_ * 0.05)
      .WithRotationalDeadband(0.1_rad_per_s)
      .WithMaxAbsRotationalRate(4.71_rad_per_s)
      .WithDriveRequestType(swerve::DriveRequestType::Velocity)
      .WithSteerRequestType(swerve::SteerRequestType::Position);
}

void RealTimeAimDrive::Execute() {
  const frc::Pose2d pose = drive_->GetState().Pose;
  const frc::Translation2d hub = drive_->GetHubPosition();
  const auto target_direction =
      frc::Rotation2d{units::math::atan2(hub.Y() - pose.Y(),
                                         hub.X() - pose.X())} -
      frc::Rotation2d{shooter_facing_offset_};
  const double angle_error =
      std::abs((target_direction - pose.Rotation()).Degrees().value());

  drive_->SOMangleDiff = angle_error;
  if (angle_error < 2.0) {
    drive_->SetControl(brake_request_);
  } else {
    drive_->SetControl(facing_request_.WithVelocityX(0_mps)
                           .WithVelocityY(0_mps)
                           .WithTargetDirection(target_direction));
  }

  frc::SmartDashboard::PutNumber("Shooting/AimErrorDeg", angle_error);
  frc::SmartDashboard::PutBoolean("Shooting/DrivetrainLocked",
                                  angle_error < 2.0);
}

void RealTimeAimDrive::End(bool interrupted) {
  drive_->SOMangleDiff = 180.0;
  const auto speeds = drive_->GetState().Speeds;
  drive_->SetControl(drive_->m_safeCoastRequest.WithVelocityX(speeds.vx)
                         .WithVelocityY(speeds.vy)
                         .WithRotationalRate(speeds.omega));
}

bool RealTimeAimDrive::IsFinished() { return false; }
